Create Association Rules for the Market Basket Analysis for the given Threshold

# Market Basket Analysis

In [1]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=d20031cbfc393ebb9698fbe2476acb6f020f61bd0f2574efeb9d629806656d00
  Stored in directory: /root/.cache/pip/wheels/77/3d/a6/d317a6fb32be58a602b1e8c6b5d6f31f79322da554cad2a5ea
Successfully built apyori


In [171]:
import pandas as pd
import numpy as np
from apyori import apriori

In [147]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [172]:
# store = pd.read_csv("/content/gdrive/MyDrive/Datasets/GroceryStoreDataSet.csv",encoding='latin-1')
store= pd.read_csv("/content/drive/MyDrive/Machine Learning /datasets/04_GroceryStoreDataSet .csv", names = ['transaction'], sep = ',')
store
store2= store

In [173]:
store.head()

transaction
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [174]:
store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 1 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   transaction  20 non-null     object
dtypes: object(1)
memory usage: 292.0+ bytes


In [175]:
#Renaming the column to PRODUCTS

store.rename(columns = {'transaction':'PRODUCTS'}, inplace = True)
store

PRODUCTS
0               MILK,BREAD,BISCUIT
1    BREAD,MILK,BISCUIT,CORNFLAKES
2              BREAD,TEA,BOURNVITA
3             JAM,MAGGI,BREAD,MILK
4                MAGGI,TEA,BISCUIT
5              BREAD,TEA,BOURNVITA
6             MAGGI,TEA,CORNFLAKES
7          MAGGI,BREAD,TEA,BISCUIT
8              JAM,MAGGI,BREAD,TEA
9                       BREAD,MILK
10  COFFEE,COCK,BISCUIT,CORNFLAKES
11  COFFEE,COCK,BISCUIT,CORNFLAKES
12          COFFEE,SUGER,BOURNVITA
13               BREAD,COFFEE,COCK
14             BREAD,SUGER,BISCUIT
15         COFFEE,SUGER,CORNFLAKES
16           BREAD,SUGER,BOURNVITA
17              BREAD,COFFEE,SUGER
18              BREAD,COFFEE,SUGER
19      TEA,MILK,COFFEE,CORNFLAKES

In [176]:
# Creating a list of items in every transaction.
#The list will work as a training set from where we can generate the list of Association Rules.

list_items = list(store["PRODUCTS"].apply(lambda x:x.split(','))) #lambda = ananymous (not named like def fname) single expression function that performs a particulatr task like map(), filter(), split(), reduce()
list_items


[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [177]:
from mlxtend.preprocessing import TransactionEncoder

encoder = TransactionEncoder()
encoded_data = encoder.fit(list_items).transform(list_items)  #silly mistake use ed
store = pd.DataFrame(encoded_data,columns = encoder.columns_)
store.head()


BISCUIT  BOURNVITA  BREAD   COCK  COFFEE  CORNFLAKES    JAM  MAGGI   MILK  \
0     True      False   True  False   False       False  False  False   True   
1     True      False   True  False   False        True  False  False   True   
2    False       True   True  False   False       False  False  False  False   
3    False      False   True  False   False       False   True   True   True   
4     True      False  False  False   False       False  False   True  False   

   SUGER    TEA  
0  False  False  
1  False  False  
2  False   True  
3  False  False  
4  False   True

In [179]:
#Applying apriori algorithm
#The lift value of an association rule is the ratio of the confidence of the rule and the expected confidence of the rule.

association_rules = apriori(list_items, min_support=0.0045, min_confidence=0.2, min_lift=3, min_length=2)
association_list = list(association_rules)

association_list

#minlength = length  of rule - 	Rules must have at least 2 items.
#minlift = 	The item on RHS is at least 3x more likely to be bought when LHS is bought.
#minconf =The rule must be correct at least 20% of the time.
#minsup = The rule must appear in at least 0.45% of transactions. (Good for small datasets)


[RelationRecord(items=frozenset({'MAGGI', 'JAM'}), support=0.1, ordered_statistics=[OrderedStatistic(items_base=frozenset({'JAM'}), items_add=frozenset({'MAGGI'}), confidence=1.0, lift=4.0), OrderedStatistic(items_base=frozenset({'MAGGI'}), items_add=frozenset({'JAM'}), confidence=0.4, lift=4.0)]),
 RelationRecord(items=frozenset({'BISCUIT', 'COFFEE', 'COCK'}), support=0.1, ordered_statistics=[OrderedStatistic(items_base=frozenset({'COCK'}), items_add=frozenset({'BISCUIT', 'COFFEE'}), confidence=0.6666666666666667, lift=6.666666666666667), OrderedStatistic(items_base=frozenset({'BISCUIT', 'COFFEE'}), items_add=frozenset({'COCK'}), confidence=1.0, lift=6.666666666666667)]),
 RelationRecord(items=frozenset({'BISCUIT', 'CORNFLAKES', 'COCK'}), support=0.1, ordered_statistics=[OrderedStatistic(items_base=frozenset({'COCK'}), items_add=frozenset({'BISCUIT', 'CORNFLAKES'}), confidence=0.6666666666666667, lift=4.4444444444444455), OrderedStatistic(items_base=frozenset({'CORNFLAKES'}), items_ad

In [182]:
#Generating the association rule between the items
#gives frwuqnt itemset

for i in range(0, len(association_list)):
    print(association_list[i][0])

frozenset({'MAGGI', 'JAM'})
frozenset({'BISCUIT', 'COFFEE', 'COCK'})
frozenset({'BISCUIT', 'CORNFLAKES', 'COCK'})
frozenset({'BISCUIT', 'COFFEE', 'CORNFLAKES'})
frozenset({'BISCUIT', 'MAGGI', 'TEA'})
frozenset({'COFFEE', 'BOURNVITA', 'SUGER'})
frozenset({'MILK', 'BREAD', 'CORNFLAKES'})
frozenset({'BREAD', 'MAGGI', 'JAM'})
frozenset({'COCK', 'COFFEE', 'CORNFLAKES'})
frozenset({'MILK', 'COFFEE', 'CORNFLAKES'})
frozenset({'COFFEE', 'TEA', 'CORNFLAKES'})
frozenset({'MILK', 'COFFEE', 'TEA'})
frozenset({'MILK', 'CORNFLAKES', 'TEA'})
frozenset({'MILK', 'MAGGI', 'JAM'})
frozenset({'TEA', 'MAGGI', 'JAM'})
frozenset({'MILK', 'BISCUIT', 'CORNFLAKES', 'BREAD'})
frozenset({'BISCUIT', 'MAGGI', 'BREAD', 'TEA'})
frozenset({'BISCUIT', 'COFFEE', 'COCK', 'CORNFLAKES'})
frozenset({'BREAD', 'MILK', 'MAGGI', 'JAM'})
frozenset({'TEA', 'BREAD', 'MAGGI', 'JAM'})
frozenset({'MILK', 'COFFEE', 'TEA', 'CORNFLAKES'})


In [183]:
from mlxtend.frequent_patterns import apriori, association_rules

frequent_itemsets = apriori(store, min_support=0.0045, use_colnames=True)
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=3)
rules


antecedents                consequents  antecedent support  \
0                 (MAGGI)                      (JAM)                0.25   
1                   (JAM)                    (MAGGI)                0.10   
2       (BISCUIT, COFFEE)                     (COCK)                0.10   
3                  (COCK)          (BISCUIT, COFFEE)                0.15   
4   (BISCUIT, CORNFLAKES)                     (COCK)                0.15   
..                    ...                        ...                 ...   
73            (MILK, TEA)       (COFFEE, CORNFLAKES)                0.05   
74     (MILK, CORNFLAKES)              (COFFEE, TEA)                0.10   
75      (CORNFLAKES, TEA)             (MILK, COFFEE)                0.10   
76                 (MILK)  (COFFEE, TEA, CORNFLAKES)                0.25   
77           (CORNFLAKES)        (MILK, COFFEE, TEA)                0.30   

    consequent support  support  confidence       lift  representativity  \
0                 0.10     0.10    0.400000   4.000000               1.0   
1                 0.25     0.10    1.000000   4.000000               1.0   
2                 0.15     0.10    1.000000   6.666667               1.0   
3                 0.10     0.10    0.666667   6.666667               1.0   
4                 0.15     0.10    0.666667   4.444444               1.0   
..                 ...      ...         ...        ...               ...   
73                0.20     0.05    1.000000   5.000000               1.0   
74                0.05     0.05    0.500000  10.000000               1.0   
75                0.05     0.05    0.500000  10.000000               1.0   
76                0.05     0.05    0.200000   4.000000               1.0   
77                0.05     0.05    0.166667   3.333333               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0     0.0750      1.5000       1.000000  0.400000   0.333333    0.700000  
1     0.0750         inf       0.833333  0.400000   1.000000    0.700000  
2     0.0850         inf       0.944444  0.666667   1.000000    0.833333  
3     0.0850      2.7000       1.000000  0.666667   0.629630    0.833333  
4     0.0775      2.5500       0.911765  0.500000   0.607843    0.666667  
..       ...         ...            ...       ...        ...         ...  
73    0.0400         inf       0.842105  0.250000   1.000000    0.625000  
74    0.0450      1.9000       1.000000  0.500000   0.473684    0.750000  
75    0.0450      1.9000       1.000000  0.500000   0.473684    0.750000  
76    0.0375      1.1875       1.000000  0.200000   0.157895    0.600000  
77    0.0350      1.1400       1.000000  0.166667   0.122807    0.583333  

[78 rows x 14 columns]

## Disp 1

In [156]:
association_rules = apriori(list_items, min_support=0.0045, min_confidence=0.2, min_lift=3, min_length=2)
association_results = list(association_rules)

# Function to visualize association rules
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))

# Creating a DataFrame of association rules
association_df = pd.DataFrame(inspect(association_results), columns = ['anticedant', 'correspondant', 'Support', 'Confidence', 'Lift'])


association_df



anticedant correspondant  Support  Confidence       Lift
0          JAM         MAGGI     0.10    1.000000   4.000000
1         COCK       BISCUIT     0.10    0.666667   6.666667
2         COCK       BISCUIT     0.10    0.666667   4.444444
3   CORNFLAKES       BISCUIT     0.10    0.333333   3.333333
4        MAGGI       BISCUIT     0.10    0.400000   4.000000
5       COFFEE         SUGER     0.05    1.000000   3.333333
6         MILK         BREAD     0.05    0.200000   4.000000
7          JAM         BREAD     0.10    1.000000   6.666667
8         COCK        COFFEE     0.10    0.666667   3.333333
9         MILK    CORNFLAKES     0.05    1.000000   3.333333
10      COFFEE    CORNFLAKES     0.05    1.000000   3.333333
11        MILK        COFFEE     0.05    0.200000   4.000000
12        MILK    CORNFLAKES     0.05    1.000000   3.333333
13         JAM          MILK     0.05    0.500000  10.000000
14       MAGGI           TEA     0.05    0.200000   4.000000
15        MILK       BISCUIT     0.05    0.200000   4.000000
16       MAGGI       BISCUIT     0.05    0.200000   4.000000
17        COCK       BISCUIT     0.10    0.666667   6.666667
18         JAM         BREAD     0.05    0.500000  10.000000
19         JAM         BREAD     0.05    0.500000   5.000000
20        MILK        COFFEE     0.05    0.200000   4.000000

## Displ

In [72]:
# Generate association rules
rules = apriori(list_items, min_support=0.0045, min_confidence=0.2, min_lift=3, min_length=2)
results = list(rules)

# Extract all rule details using a for loop
data = []

for rule in results:
    for i in rule[2]:
        antecedent = list(i[0])[0] if len(i[0]) == 1 else tuple(i[0])
        consequent = list(i[1])[0] if len(i[1]) == 1 else tuple(i[1])
        support = rule[1]
        confidence = i[2]
        lift = i[3]
        data.append((antecedent, consequent, support, confidence, lift))

# Create DataFrame
df_rules = pd.DataFrame(data, columns=['Antecedent', 'Consequent', 'Support', 'Confidence', 'Lift'])

df_rules


Antecedent             Consequent  Support  Confidence  \
0                         JAM                  MAGGI     0.10    1.000000   
1                       MAGGI                    JAM     0.10    0.400000   
2                        COCK      (BISCUIT, COFFEE)     0.10    0.666667   
3           (BISCUIT, COFFEE)                   COCK     0.10    1.000000   
4                        COCK  (BISCUIT, CORNFLAKES)     0.10    0.666667   
..                        ...                    ...      ...         ...   
68         (MILK, CORNFLAKES)          (COFFEE, TEA)     0.05    0.500000   
69          (CORNFLAKES, TEA)         (COFFEE, MILK)     0.05    0.500000   
70                (MILK, TEA)   (COFFEE, CORNFLAKES)     0.05    1.000000   
71  (COFFEE, TEA, CORNFLAKES)                   MILK     0.05    1.000000   
72        (MILK, COFFEE, TEA)             CORNFLAKES     0.05    1.000000   

         Lift  
0    4.000000  
1    4.000000  
2    6.666667  
3    6.666667  
4    4.444444  
..        ...  
68  10.000000  
69  10.000000  
70   5.000000  
71   4.000000  
72   3.333333  

[73 rows x 5 columns]

## AP 2

In [ ]:
#####  APRIORI 2

In [118]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules
store = pd.read_csv("/content/drive/MyDrive/Machine Learning /datasets/04_GroceryStoreDataSet .csv", names = ['transaction'], sep = ',')
store

transaction
0               MILK,BREAD,BISCUIT
1    BREAD,MILK,BISCUIT,CORNFLAKES
2              BREAD,TEA,BOURNVITA
3             JAM,MAGGI,BREAD,MILK
4                MAGGI,TEA,BISCUIT
5              BREAD,TEA,BOURNVITA
6             MAGGI,TEA,CORNFLAKES
7          MAGGI,BREAD,TEA,BISCUIT
8              JAM,MAGGI,BREAD,TEA
9                       BREAD,MILK
10  COFFEE,COCK,BISCUIT,CORNFLAKES
11  COFFEE,COCK,BISCUIT,CORNFLAKES
12          COFFEE,SUGER,BOURNVITA
13               BREAD,COFFEE,COCK
14             BREAD,SUGER,BISCUIT
15         COFFEE,SUGER,CORNFLAKES
16           BREAD,SUGER,BOURNVITA
17              BREAD,COFFEE,SUGER
18              BREAD,COFFEE,SUGER
19      TEA,MILK,COFFEE,CORNFLAKES

pandas for dataframe management
mlxtend (machine learning extensions) is a Python library of useful tools for day-to-day data science tasks.

mlxtend.preprocessing for data preprocessing before performing the task.

mlxtend.frequent_patterns for association rule task.

load data as a variable df

In [119]:
#2. One-hot encoding
#We have to change the data to the appropriate format before inputting it into the Apriori algorithm. The goal is a one-hot DataFrame.

#Change the DataFrame to a list of lists.

store2 = list(store["transaction"].apply(lambda x:x.split(",")))
store2


[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [86]:
#df = list(df["transaction"].apply(lambda x:x.split(",")))

In [28]:
#Use TransactionEncoder from the mlxtend library to change the list of transactions to a one-hot array.

In [120]:
one_hot_transformer = TransactionEncoder()
df_transform = one_hot_transformer.fit_transform(store2)

In [121]:
#Change the one-hot array to a DataFrame.
df = pd.DataFrame(df_transform,columns=one_hot_transformer.columns_)

This is our desired format, one-hot. The columns are items in the store and each row represents a transaction. If the value is True, that item is sold in that transaction. Now, the data is ready to be fed to the algorithm

In [122]:
#3. Find the frequent itemsets using Apriori

df = apriori(df, min_support = 0.2, use_colnames = True)
df.sort_values(['support'],ascending=False, inplace = True)


# df_ar = association_rules(df, metric="lift", min_threshold=1)
# df_ar

 the minimum support to 0.2, so it will filter only a set of item(s) whose support is greater than 0.2 (or more than 20% from 19 transactions).

Only these itemsets which are considered important will proceed to the association rule.

In [123]:
#4. Association rule

df_ar = association_rules(df, metric="lift", min_threshold=1)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0       (BREAD)        (MILK)                0.65                0.25   
1        (MILK)       (BREAD)                0.25                0.65   
2       (BREAD)       (SUGER)                0.65                0.30   
3       (SUGER)       (BREAD)                0.30                0.65   
4      (COFFEE)  (CORNFLAKES)                0.40                0.30   
5  (CORNFLAKES)      (COFFEE)                0.30                0.40   
6      (COFFEE)       (SUGER)                0.40                0.30   
7       (SUGER)      (COFFEE)                0.30                0.40   
8       (MAGGI)         (TEA)                0.25                0.35   
9         (TEA)       (MAGGI)                0.35                0.25   

   support  confidence      lift  representativity  leverage  conviction  \
0      0.2    0.307692  1.230769               1.0    0.0375    1.083333   
1      0.2    0.800000  1.230769               1.0    0.0375    1.750000   
2      0.2    0.307692  1.025641               1.0    0.0050    1.011111   
3      0.2    0.666667  1.025641               1.0    0.0050    1.050000   
4      0.2    0.500000  1.666667               1.0    0.0800    1.400000   
5      0.2    0.666667  1.666667               1.0    0.0800    1.800000   
6      0.2    0.500000  1.666667               1.0    0.0800    1.400000   
7      0.2    0.666667  1.666667               1.0    0.0800    1.800000   
8      0.2    0.800000  2.285714               1.0    0.1125    3.250000   
9      0.2    0.571429  2.285714               1.0    0.1125    1.750000   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.535714  0.285714   0.076923    0.553846  
1       0.250000  0.285714   0.428571    0.553846  
2       0.071429  0.266667   0.010989    0.487179  
3       0.035714  0.266667   0.047619    0.487179  
4       0.666667  0.400000   0.285714    0.583333  
5       0.571429  0.400000   0.444444    0.583333  
6       0.666667  0.400000   0.285714    0.583333  
7       0.571429  0.400000   0.444444    0.583333  
8       0.750000  0.500000   0.692308    0.685714  
9       0.865385  0.500000   0.428571    0.685714

The association_rules function will automatically calculate key metrics of our transaction data including support, confidence, lift, leverage, and conviction.

The result is required interpretation for further business actions.

Interpretation examples

From the highest confidence in index 0, the confidence is 0.8. It means that the customer who buys milk will buy bread for 80%. However, you have to keep in mind that confidence is not everything. The high confidence in this row is due to the high support of bread (consequence support 0.65) which means bread occurs in many transactions, so it will not make a business impact if we try to sell bread with milk anyway.
The better metric is lift or leverage. Index 8 has the highest lift and leverage at 2.28 and 0.1125 respectively. It means that the customers who buy tea are likely to buy Maggi as well.
What we can do after knowing this?

Change shelf layout to sell Maggi close to tea.
Promoting Maggi together with tea.